prow/metrics/metrics_test.go

```diff
@@ -0,0 +1,143 @@
+		newCh := make(chan time.Time, 1)
+		newCh <- time.Time{}
+		waitForAfter <- d
+		return newCh
+	}
+
+	pusher := Pusher{
+		waiter:   fakeAfter,
+		gatherer: nullGatherer,
+		configAgent: &fakeConfigAgent{
+			fakeConfigs: []config.PushGateway{
+				// Initial wait
+				{Interval: 10 * time.Second},
+				{Interval: 20 * time.Minute},
+				// This interval is smaller then the min allowed duration
+				{Interval: 0},
+				{Interval: 30 * time.Second},
+				// This interval is bigger then the max allewd duration
+				{Interval: 30 * time.Hour},
+				{Interval: 10 * time.Second},
+			},
+		},
+	}
+	go pusher.Start("")
+
+	expectedIntervals := []time.Duration{
+		10 * time.Second,
+		20 * time.Minute,
+		minWaitDuration,
+		30 * time.Second,
+		maxWaitDuration,
+	}
+	for i, expected := range expectedIntervals {
+		if actual := <-waitForAfter; actual != expected {
+			t.Fatalf("Expected interval %d to be '%s', got: '%s'", i, expected, actual)
+		}
+	}
+}
+
+// TestMetricsPusherGathererCalls tests that ...
+// - the pusher does not run the gatherer function if the endpoint is not
+//   configured (setting the endpoint to "" can be used to temporatily disable
+//   the pushing/gathering of metrics)
+// - keeps the pusher loop running, regardless if endpoint is configured or not
+// - if the endpoint is set again, the gatherer function will be called again
+func TestMetricsPusherGathererCalls(t *testing.T) {
+	waitForURL := make(chan string)
+
+	fakeGatherer := func(job string, grouping map[string]string, url string, g prometheus.Gatherer) error {
+		waitForURL <- url
+		return nil
+	}
+
+	pusher := Pusher{
+		gatherer: fakeGatherer,
+		waiter:   nullWaiter,
+		configAgent: &fakeConfigAgent{
+			fakeConfigs: []config.PushGateway{
+				// The first time the Config() is requested is just to get the initial
+				// sleep interval, subsequent calls will use both the interval and
+				// the endpoint
+				{Endpoint: "will be ignored"},
+				{Endpoint: ""},
+				{Endpoint: "endpoint one"},
+				{Endpoint: ""},
+				{Endpoint: ""},
+				{Endpoint: "endpoint two"},
+				{Endpoint: ""},
+				{Endpoint: "endpoint three"},
+			},
+		},
+	}
+	go pusher.Start("")
+
+	expectedEndpoints := []string{
+		"endpoint one",
+		"endpoint two",
+		"endpoint three",
+	}
+	for i, expected := range expectedEndpoints {
+		if actual := <-waitForURL; actual != expected {
+			t.Fatalf("Expected endpoint %d to be '%s', got: '%s'", i, expected, actual)
+		}
+	}
+}
+
+func nullWaiter(wait time.Duration) <-chan time.Time {
+	ch := make(chan time.Time, 1)
+	ch <- time.Time{}
+	return ch
+}
+
+func nullGatherer(job string, grouping map[string]string, url string, g prometheus.Gatherer) error {
+	return nil
+}
+
+type fakeConfigAgent struct {
+	fakeConfigs []config.PushGateway
+	idx         int
+}
+
+func (ca *fakeConfigAgent) Config() *config.Config {
+
+	c := &config.Config{}
+	c.PushGateway = ca.fakeConfigs[ca.idx]
+
+	if ca.idx < len(ca.fakeConfigs)-1 {
```

Comment:
It's strange that the fake config agent is different from the real config agent. Instead of using an index internal to the fake agent, I would prefer to see table-driven unit tests that reconstruct the config agent every time. It's going to be much simpler to follow such unit tests.

Reply:
A `ConfigAgent` might return a different config each time `Config()` is called. This is important for `TestMetricsPusherGathererCalls` as depending on the value of `Endpoint` it might actually call the `gatherer` or not.
So the `fakeConfigAgent` is setup in a way to return different configs, in this case by returning on config from the list of pre-configured configs (mimicking config updates) or the last config of the list when all other have already been returned (mimicking no config update). Therefore using a new `ConfigAgent` with a single config to return is not enough here IMHO.
Also, we think the `fakeConfigAgent` is pretty close to the actual `ConfigAgent` (potentially returning changed config on consecutive calls) -- but obviously it is not that clear as we thought. I thought about how this can be made clearer and tested in a better way but couldn't come up with something better yet. 

A table test with a newly created simple `ConfigAgent` would work for `TestMetricsPusherWaiterCalls` as we expect the `waiter` to be called no matter what. We are happy to change that I you think that makes more sense. I'd slightly lean towards keeping that test as is (= the same test setup as the other test) but I have no hard opinion about that.